apiserver/routers/routers.go

```diff
@@ -23,13 +23,20 @@ import (
 	"garm/apiserver/controllers"
 	"garm/auth"
 	"garm/util"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
 )
 
 func NewAPIRouter(han *controllers.APIController, logWriter io.Writer, authMiddleware, initMiddleware, instanceMiddleware auth.Middleware) *mux.Router {
 	router := mux.NewRouter()
 	logMiddleware := util.NewLoggingMiddleware(logWriter)
 	router.Use(logMiddleware)
 
+	// Handles github webhooks
+	router.Handle("/metrics", promhttp.Handler())
```

Comment:
I think we should create a sub router for the metrics endpoint. We could then apply any kind of middleware we want just on that router. 

Something like:

```suggestion
	metricsRouter := router.PathPrefix("/metrics").Subrouter()
	metricsRouter.Handle("/", promhttp.Handler())
```

Which brings me to the authentication side of things. Can we add any kind of authentication? Even basic auth would do. Prometheus seems to allow the ```Authorization``` header to be set in the ```scrape_config``` section, which means we can use the regular ```JWT``` token auth. Or at least basic auth.

Not very comfortable exposing metrics publicly.

Reply:
I was thinking of creating a new auth middleware for metrics, limited in scope, just like the instance callback middleware. We could add an API endpoint where we could generate tokens for the metrics service, and you could do something like:

```bash
garm-cli metrics token create
```

With optional parameters to limit the validity time, and possibly other context fields, which you could then configure in your Prometheus ```scrape_config```.

These tokens would only work against ```/metrics```. The ```/metrics``` endpoint would be accessible to anyone with an ```admin``` context or the limited ```metrics``` token. This way we could possibly (later) expose the metrics to the garm CLI as well.

I can have a stab at this if you'd like. Would mandatory auth be an issue for your setup?

Reply:
As a general thing, I tend to treat other systems as adversarial. Even if they are in the same network and even if there are some firewall rules in place to limit which systems can talk to other systems. If there is any chance of exposing any kind of sensitive information, I usually do it after checking that the remote system was granted explicit access to do so.

Reply:
that is fine for me .. but probably you'll have to take a lead in implementing this.

Still, I think it should be possible to disable metrics auth. Normally, e.g. in a k8s cluster, some (probably managebd by somebody else) prometheus automatically discovers 'things' to scrape. And this would be impossible if it needs to authenticate itself.

Reply:
We can have it without auth to start with, but I think we should document this [in the docs folder](/cloudbase/garm/tree/main/doc), and also amend the sample ```nginx``` config file to set up a separate ```location /metrics``` as an example of how people could put this behind a reverse proxy and potentially add basic auth using nginx.

What do you think?

Reply:
Apropos, the [prometheus scape_config](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#%3Cscrape_config%3E) has the [tls_config](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#tls_config) section which allows you to set the client certificates.

There is an example here:

https://github.com/prometheus/prometheus/blob/main/documentation/examples/prometheus-kubernetes.yml

Reply:
.. it's possible in prometheus, but if this prometheus is managed by a 3rd party and scans (a potentially big) number of services this 3rd party would have to handle (store, delete, update) many customer secrets. 
I think in reality in such a setup prometheus scrapes only endpoints that have no own auth but are not exposed anywhere. prometheus then uses kubeproxy to discover and scrape them.

Shall we start exposing metrics on an own port? and try to offer optional auth for that in a 2nd step ?

Reply:
I think we can get away with putting it behind a reverse proxy like nginx for now. In nginx we can define a ```location /metrics``` that proxy passes back to garm, and in that ```location``` block if anyone wants to, can enable basic auth or whatever other kind of restrictions.

I can have a look at creating a middleware for the ```/metrics``` endpoint next week.